workflow/controller/operator_test.go

```diff
@@ -2106,3 +2106,109 @@ func findNodeByName(nodes map[string]wfv1.NodeStatus, name string) *wfv1.NodeSta
 	}
 	return nil
 }
+
+var invalidSpec = `
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  name: invalid-spec
+spec:
+  entrypoint: 123
+`
+
+func TestEventInvalidSpec(t *testing.T) {
+	controller := newController()
+	wfcset := controller.wfclientset.ArgoprojV1alpha1().Workflows("")
+	wf := unmarshalWF(invalidSpec)
+	wf, err := wfcset.Create(wf)
+	assert.NoError(t, err)
+	woc := newWorkflowOperationCtx(wf, controller)
+	woc.operate()
+	events, err := controller.kubeclientset.CoreV1().Events("").List(metav1.ListOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(events.Items))
+	runningEvent := events.Items[0]
+	assert.Equal(t, "WorkflowRunning", runningEvent.Reason)
+	invalidSpecEvent := events.Items[1]
+	assert.Equal(t, "WorkflowFailed", invalidSpecEvent.Reason)
+	assert.Equal(t, "invalid spec: template name '123' undefined", invalidSpecEvent.Message)
+}
+
+var timeout = `
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  name: timeout-template
+spec:
+  entrypoint: sleep
+  activeDeadlineSeconds: 1
+  templates:
+  - name: sleep
+    container:
+      image: alpine:latest
+      command: [sh, -c, sleep 10]
+`
+
+func TestEventTimeout(t *testing.T) {
```

Comment:
please add a comment to explain what the test is for

Reply:
done